src/geoip_cache.rs

```diff
@@ -0,0 +1,115 @@
+use serde::{Deserialize, Serialize};
+use std::net::IpAddr;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+use std::time::SystemTime;
+use std::{collections::HashMap, fs, io, time::Duration};
+use tokio::sync::RwLock;
+
+use ziggurat_core_geoip::geoip::{GeoIPInfo, GeoIPService};
+
+#[derive(Clone, Serialize, Deserialize)]
+struct CacheEntry {
+    pub last_updated: SystemTime,
+    pub entry: GeoIPInfo,
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+struct GeoCache {
+    pub entries: HashMap<IpAddr, CacheEntry>,
+}
+
+/// GeoIP cache responsible for getting and caching results.
+pub struct GeoIPCache {
+    /// Available providers and their configuration.
+    providers: Vec<Box<dyn GeoIPService>>,
+    /// Path to the cache file.
+    cache_file: PathBuf,
+    /// Cache entries.
+    cache: Arc<RwLock<GeoCache>>,
+}
+
+impl GeoIPCache {
+    /// Create a new GeoIP cache.
+    pub fn new(cache_file: &Path) -> Self {
+        Self {
+            providers: Vec::new(),
+            cache_file: cache_file.to_owned(),
+            cache: Arc::new(RwLock::new(GeoCache::new())),
+        }
+    }
+
+    /// Add a new provider to the list of providers. The providers will be called in the order they
+    /// are added.
+    pub fn add_provider(&mut self, provider: Box<dyn GeoIPService>) {
+        self.providers.push(provider);
+    }
+
+    /// Load the cache from the file.
+    pub async fn load(&self) -> Result<(), io::Error> {
+        let cache_string = fs::read_to_string(&self.cache_file)?;
+
+        let mut cache = self.cache.write().await;
+        cache.entries = serde_json::from_str(&cache_string).unwrap();
+        Ok(())
+    }
+
+    /// Save the cache to the file.
+    pub async fn save(&self) -> Result<(), io::Error> {
+        let cache = self.cache.read().await;
+        let cache_string = serde_json::to_string(&cache.entries).unwrap();
+        fs::write(&self.cache_file, cache_string)
+    }
+
+    /// Function look in cache and if not found, it will call the providers to fetch new data and
+    /// store it into cache.
+    pub async fn lookup(&self, ip: IpAddr) -> Option<GeoIPInfo> {
+        let mut remove_entry = false;
+
+        {
+            let cache = self.cache.read().await;
+            let res = cache.entries.get(&ip);
+            if let Some(entry) = res {
+                // Check if the entry is not too old.
+                if entry.last_updated.elapsed().unwrap() < Duration::from_secs(60 * 60 * 24 * 14) {
+                    return Some(entry.entry.clone());
+                } else {
+                    remove_entry = true;
+                }
+            }
+        }
+
+        {
+            // Beeing here means that we didn't find the entry in cache.
+            let mut rw_cache = self.cache.write().await;
+
+            if remove_entry {
```

Comment:
Can we move locking to inside `if`, then take lock below only if `let Ok(entry)` is true?

Reply:
We could but we need also write lock to remove outdated entry.
There is possible that we'd remove outdated entries only when new data is fetched properly (so move `remove_entry` `if` inside `let Ok(entry)` but it could potentially lead to the situation when there will be many outdated entries kept in the cache.
Personally, I don't believe with such situation and I can move removing entries to the place where we've new data for them (well, in such case we can just modify it). As a tradeoff - some entries can be outdated and still kept in cache but they should not be a big deal.

Reply:
> We could but we need also write lock to remove outdated entry.

Yes, that's what I mean. We would take the lock twice: once before removal (inside if remove_entry), and the second before insert (inside if let Ok(entry)...). My reasoning is because provider.lookup can take a long time, right?

Reply:
Yea, it could be for some reasons. Changed! :)